log_mix - Multivariate Containers

andrjohns opened the pull request:
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
This pull extends log_mix to take a vector of mixing proportions and an array of density vectors (i.e. ```log_mix(vector, vectors[])``` with analytic gradients. Any combination of vector/row_vector/std::vector is valid.

This required extending operands_and_partials to work with the ```std::vector<std::vector<T>>``` signature, let me know if that would be better introduced in a separate pull.

The tests should be fairly thorough and cover every possible combination of inputs, but let me know if I've missed anything.

#### Intended Effect:
Improve the speed and efficiency of mixture modelling

#### How to Verify:
Tests for ```prim/fwd/rev/mix``` are included

#### Side Effects:
N/A

#### Documentation:
As doxygen

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Andrew Johnson

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

syclik commented:
@seantalts, could you review this one? It touches operands and partials.

seantalts commented:
Sure thing. I need to read up on what log_mix should be doing and whatnot so it might take me a little while but I'm happy to learn.

bob-carpenter commented:
For binary mixtures, it's

log_mix(real<lower=0, upper=1> alpha, real lp1, real lp2) 

=def=

log_sum_exp(log(alpha) + lp1, log1m(alpha) + lp2)

That can be vectorized in the usual way.


For general mixtures, it's

log_mix(simplex theta, vector lp)

=def=

log_sum_exp(log(theta) + lp)

andrjohns commented:
That's right, the general mixtures form was introduced in #664 and this pull extends that to:

log_mix(simplex theta, array of vectors{lp1,...,lpN})

=def=

log_sum_exp(log(theta) + lp1) + ... + log_sum_exp(log(theta) + lpN)

bob-carpenter commented:
Neat---vectorizing the vector form.

seantalts commented:
Hm, the manual says this: "There is (currently) no way to vectorize mixture models at the observation level in Stan. This section is to warn users away from attempting to vectorize naively, as it re- sults in a different model."

bob-carpenter commented:
This mixes at the observation level:
```
for (n in 1:N)
  target += log_mix(theta, normal_lpdf(y[n] | mu[1], sigma[1]), normal_lpdf(y[n] | mu[2], sigma[2]));
```
whereas this mixes at the entire data set level:
```
target += log_mix(theta, normal_lpdf(y | mu[1], sigma[1]), normal_lpdf(y | mu[2], sigma[2]));
```
There's no way to vectorize the former.  There still won't be after this pull request, because there's no way to get `normal_lpdf` to return a vector.

syclik commented:
@andrjohns, it looks like there are new warnings in this branch. Just based on a quick look, it seems like you're not using the correct type for indexing somewhere.

andrjohns commented:
Ah, rookie mistake. Looks like it's sorted now, thanks for the heads up

andrjohns commented:
@seantalts Is there a way to see more output on Jenkins? This failed the Stan downstream tests and I'm not sure why

seantalts commented:
You can use the little door in the top right to exit to the old view and then you should be able to just click on the Stan build it kicked off.

andrjohns commented:
Thanks for that, right in front of me. I've had a look over the logs for each stage, and the pull is actually passing the Stan downstream unit tests, but there's some failure in the "Archive JUnit-formatted test results" and "Recursively delete the current directory from the workspace" steps. I can't see any actual errors in the output, so I'm not entirely sure what's causing the failures.

sakrejda commented:
I couldn't find anything wrong (I did click through the Stan downstream test but it wouldn't show me the logs for the failed step).  Nothing else was running so I clicked the 'restart' for that test in case that's all it needed.

seantalts commented:
The mac pro (gelman-group-mac) has been dropping network connections a lot
in the past week or two. I haven't figured out what's causing it - likely
some new bug in Jenkins, unless you were adding something particularly
taxing to the compiler with this PR.

On Wed, Mar 7, 2018 at 11:41 AM, Krzysztof Sakrejda <
notifications@github.com> wrote:

> I couldn't find anything wrong (I did click through the Stan downstream
> test but it wouldn't show me the logs for the failed step). Nothing else
> was running so I clicked the 'restart' for that test in case that's all it
> needed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/751#issuecomment-371200388>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7PGhXjWggHxrFXRTklqm8WOSUk1kks5tcA2hgaJpZM4R7nuw>
> .
>

andrjohns commented:
That seems to have done the trick, thanks

syclik commented:
@seantalts, still up for reviewing this PR?

syclik commented:
If not, please unassign yourself and assign me.

seantalts commented:
At one point I thought Bob had said he would, but I might have
misunderstood. I will take a look tomorrow, might need backup if there’s
tricky math.
On Mon, Mar 12, 2018 at 22:38 Daniel Lee <notifications@github.com> wrote:

> If not, please unassign yourself and assign me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/751#issuecomment-372527209>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7Ma9yfh6rkNP1vNf-aog1V1DAmANks5tdzEUgaJpZM4R7nuw>
> .
>

bob-carpenter commented:
Nope, shouldn't be any tricky math there.

- Bob

> On Mar 12, 2018, at 10:52 PM, seantalts <notifications@github.com> wrote:
> 
> At one point I thought Bob had said he would, but I might have
> misunderstood. I will take a look tomorrow, might need backup if there’s
> tricky math.
> On Mon, Mar 12, 2018 at 22:38 Daniel Lee <notifications@github.com> wrote:
> 
> > If not, please unassign yourself and assign me.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/stan-dev/math/pull/751#issuecomment-372527209>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AAxJ7Ma9yfh6rkNP1vNf-aog1V1DAmANks5tdzEUgaJpZM4R7nuw>
> > .
> >
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

seantalts commented on stan/math/prim/mat/fun/log_mix.hpp:
```diff
@@ -94,6 +97,274 @@ typename return_type<T_theta, T_lam>::type log_mix(const T_theta& theta,
+
+  T_partials_mat derivs
+      = (lam_dbl - logp.transpose().replicate(M, 1))
+            .unaryExpr([](T_partials_return x) { return exp(x); });
```
This seems less efficient than doing the exp first and then replicating, no?

andrjohns replied:
The replicate is only applied to the vector ```logp``` so that ```lam_dbl``` and ```logp``` are matrices of identical dimensions and the subtraction can be done in a single step (essentially a row-wise subtraction of a vector from a matrix). The exp operation is then applied to each element of the resulting matrix.
i.e.: ```exp(lam_dbl - logp.transpose().replicate(M, 1))```

seantalts replied:
Right - isn't this more efficient:
```
exp(lam_dbl - logp.transpose()).replicate(M, 1)
```
(notice the placement of the parens)

seantalts replied:
Hmm, seems like I'm wrong about that. I guess replicate does some special sauce under the hood. The weirdest thing is that my benchmark actually thinks the first version (exp'ing the replicate instead of replicating the exp'd) [is way faster](https://github.com/seantalts/cppperf/blob/master/eigen_replicate_results.txt)

seantalts replied:
Wait, no, I got mixed up in the results. "ReplicateAfter" is way faster, i.e. we should call exp first, i.e. [the 2nd version](https://github.com/stan-dev/math/pull/751#discussion_r176154760) with the paren moved inwards is better.

andrjohns replied:
But that's not a valid operation? ```lam_dbl``` is a matrix with dimensions [M,N] and ```logp``` is a vector with dimensions [N,1] (making ```logp.transpose()``` a row vector with dimensions [1,N].

```([M,N] - [1,N])``` is not valid arithmetic in eigen (https://eigen.tuxfamily.org/dox/group__TutorialMatrixArithmetic.html): 
>Addition and subtraction
>The left hand side and right hand side must, of course, have the same numbers of rows and of columns

The ```replicate()``` is needed so that ```logp.transpose()``` has dimensions [M,N] and the resulting operation is ```([M,N] - [M,N])```

Attempting to evaluate ```exp(lam_dbl - logp.transpose()).replicate(M, 1) ``` returns:
```
Assertion `(aLhs.rows() == aRhs.rows() && aLhs.cols() == aRhs.cols())` failed.
Aborted (core dumped)
```

seantalts replied:
Oh, you're right, didn't notice the difference in dimensionality. Sorry for the confusion.

seantalts commented on stan/math/prim/mat/fun/log_mix.hpp:
```diff
@@ -94,6 +97,274 @@ typename return_type<T_theta, T_lam>::type log_mix(const T_theta& theta,
+  for (int n = 0; n < N; ++n)
+    logp[n] = log_sum_exp(logp_tmp.col(n).eval());
+
+  T_partials_mat derivs
```
Seems like we only want to calculate the derivatives if we need them - can we put this and the following ops_partials math inside something like `if (!is_constant_struct<T_theta>::value || !is_constant_struct<T_lam>::value)`?

andrjohns replied:
That would be a good idea. The matrix ```derivs``` is manipulated in different ways for each input, so would just forgo the calculations if neither input needed derivatives.

i.e.  ```(!(is_constant_struct<T_theta>::value && is_constant_struct<T_lam>::value))```

seantalts commented on stan/math/prim/mat/fun/log_mix.hpp:
```diff
@@ -94,6 +97,274 @@ typename return_type<T_theta, T_lam>::type log_mix(const T_theta& theta,
+template <typename T_theta, typename T_lam>
+typename return_type<T_theta, std::vector<Eigen::Matrix<T_lam, 1, -1> > >::type
+log_mix(const T_theta& theta,
+        const std::vector<Eigen::Matrix<T_lam, 1, -1> >& lambda) {
```
The comments above apply here, which makes me wonder if we can't share some of this code?

andrjohns replied:
I've no doubt that it could be reduced, but the difficulty is that I'm not sure how. This function is a bit of a pet project for me, so I'll be working on optimising it as I learn more about template programming.

andrjohns commented on stan/math/prim/mat/meta/operands_and_partials.hpp:
```diff
@@ -15,14 +15,14 @@ namespace internal {
  */
-template <int R, int C>
-class ops_partials_edge<double, Eigen::Matrix<double, R, C>> {
+template <typename T_op, typename T_part, int R, int C>
```
The templates needed to be changed to allow for cases where the input simplex was a vector of ```fvar<var>```, and the input densities were an array of vectors of doubles (i.e. ```log_mix(Matrix<fvar<var>, Dynamic, 1>, std::vector<Matrix<double, Dynamic, 1>>```)

This particular template (```ops_partials_edge<double, Eigen::Matrix<double, R, C>>```) did not need to be changed, but I thought it was better to maintain consistency with the others

seantalts replied:
The operands_and_partials in this file and directory are for primitive types, and this isn't just a code organizational concern. The `dx` methods won't work if you're hoping these template specializations will capture and deal with types containing `<fvar<var>` appropriately.

seantalts replied:
If I'm understanding the situation correctly, we'll need to change the code in `fwd/mat/meta/operands_and_partials` to deal with these types of nested Eigen matrices of `fvar<Dx>`s.

andrjohns replied:
The changes to ```fwd/mat/meta/operands_and_partials``` are also included in [the pull](https://github.com/stan-dev/math/pull/751/files#diff-7f6700f663c921cc2ba73d3b5175ad10).

> The ```dx``` methods won't work if you're hoping these template specializations will capture and deal with types containing ```fvar<var>``` appropriately.

There's a very good chance that I'm not fully understanding this, but the tests with```fvar<var>``` inputs all return the correct values? In the ```mix``` tests [here](https://github.com/andrjohns/math/blob/5767f8d6392c6a0c2c0494051126b9a74ad865a3/test/unit/math/mix/mat/fun/log_mix_test.cpp#L16)

seantalts replied:
Hmm, I see that you have one new specialization in the PR for `operands_and_partials` for `std::vector<std::vector<fvar<Dx>>`. I was saying these matrix specializations that you changed in `prim/mat` from `double` to `T_op` that were intended to deal with both matrices of doubles and matrices of fvars might need to be split out instead into two versions, though it was a very good instinct to try to unify the code. But we actually want a version that does return 0 for all inputs here in `prim/mat` because we don't have derivatives w.r.t. `double` operands. It's interesting that your tests pass, because none of these operands_and_partials types in `prim/mat` have `dx()` implemented in such a way as to return anything other than 0 for all inputs... So I'm pretty curious what's happening here. Perhaps another specialization ends up covering these cases, if we are getting the right derivatives and everything for Eigen types?

Could you add new tests for these Eigen types for `fwd/mat/operands_and_partials`? Looks like the one we have now just does a single `std::vector`. It's a bit ironic, because we don't use forward mode (at all; obviously we don't use the new functionality you're adding here), but we ask that all modes are tested in the hopes that forward mode is one day fully tested and we can use it for new forms of HMC and optimization.


EDIT: I just looked it up - we do actually have specializations in `fwd/mat` for Eigen types `fvar<Dx>`. Did you realize that and were trying to add something new that I'm still misunderstanding? I'm sorry about the additional confusion, if so. Here is where I'm looking: https://github.com/stan-dev/math/blob/develop/stan/math/fwd/mat/meta/operands_and_partials.hpp#L40

andrjohns replied:
Ah, I think that this is actually my misunderstanding. Should ```mix``` tests for functions that take two inputs be testing combinations of ```double``` and ```fvar<var>```?

i.e. should the mix tests include ```foo(fvar<var>, double)``` and ```foo(fvar<fvar<var>>, double)``` or just ```foo(fvar<var>,fvar<var>)``` and ```foo(fvar<fvar<var>>, fvar<fvar<var>>)```?

seantalts replied:
Ideally, yeah.

andrjohns replied:
Ok, I'm starting to drill down to the problem.

Operating under the ```develop``` branch, if a function calls:

```
  fvar<var> a(2.0, 2.0);
  double b = 2.0;

  operands_and_partials<fvar<var>, double> o(a, b);
  o.edge1_.partials_[0] = 5.0;
  if (!is_constant_struct<double>::value)
    o.edge2_.partials_[0] = 5.0;
```

```o.edge2_``` is correctly instantiated under the ```prim/scal/operands_and_partials``` template, since it allows the operand (```0p```) and the partial (```ViewElt```) to be different types.

On the other hand, if a function calls:
```
  fvar<var> a(2.0, 2.0);

  Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1> avec(3);
  avec << a, a, a;

  Eigen::Matrix<double, Eigen::Dynamic, 1> bvec(3);
  bvec << 2.0, 2.0, 2.0;

  Eigen::Matrix<double, Eigen::Dynamic, 1> dxvec(3);
  dxvec << 5.0, 5.0, 5.0;

  operands_and_partials<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>, 
                        Eigen::Matrix<double, Eigen::Dynamic, 1>> o(avec, bvec);
  o.edge1_.partials_vec_[0] = dxvec;
  if (!is_constant_struct<Eigen::Matrix<double, Eigen::Dynamic, 1>>::value)
    o.edge2_.partials_vec_[0] = dxvec;
```

```o.edge2_``` does not get instantiated under the ```prim/mat/operands_and_partials``` template because the template expects both the operand and the partial to be of type double.

So the templates between ```prim/scal``` and ```prim/mat``` are actually different in the combinations of types that they allow.

seantalts replied:
That failure in operands_and_partials makes sense to me - we shouldn't have partial derivatives for a matrix of doubles. The derivative with respect to a constant is always 0. I think the problem might lie in `is_constant_struct` not understanding that a Matrix of doubles is constant?

seantalts replied:
Quickly reading the code, it seems like I would expect `is_constant_struct` to work there, and [the tests](https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/mat/meta/is_constant_struct_test.cpp#L9) seem to confirm this... How is that code executing?

seantalts replied:
Oh wait, sorry, I'm just waking up. The compiler still needs that to compile, even if it statically evaluates to false and is never called at runtime. It looks like in your PR, the correct `prim/mat/meta/operands_and_partials` templates with a `partials_vec_` member are no longer being instantiated appropriately. I suspect if you revert your changes to `prim/mat/meta/operands_and_partials.hpp` that this will start resolving to the correct template again.

andrjohns replied:
This occurs under develop as well, not just my PR

seantalts replied:
Ah, okay. Can you add a test for the proposed new behavior in operands_and_partials (required anyway) and I can dig into it this weekend or next week? This would be a new test file in `prim/mat/meta/operands_and_partials_test.cpp` showcases the things that should now work and things that shouldn't work. Thanks!

seantalts replied:
I had a few minutes and am investigating further, driven partially by curiosity, and I did not notice that for some reason operands_and_partials thinks that the ViewElt of an eigen matrix is a `var`. That is where the problem stems from, I believe.

seantalts replied:
I think we actually need to change the code here: https://github.com/stan-dev/math/blob/develop/stan/math/fwd/scal/meta/operands_and_partials.hpp#L63 to not pass Dx but actually whatever the ViewElt would be for Op2, 3, etc.

I will look more into it later. Having ViewElt = var isn't right..

andrjohns replied:
Progress is good! Unfortunately your changes break the usage of ```partials_return_type```.

If changing the test case posted above to:
```
#include <stan/math/mix/mat.hpp>
#include <stan/math/prim/scal/meta/is_constant_struct.hpp>
#include <gtest/gtest.h>

TEST(MetaTraits, partials_type) {
  using stan::math::fvar;
  using stan::math::var;
  using stan::is_constant_struct;
  using stan::math::operands_and_partials;

  typedef typename stan::partials_return_type<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>,
                                              Eigen::Matrix<double, Eigen::Dynamic, 1>>::type
      T_partials_return;

  fvar<var> a(2.0, 2.0);

  Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1> avec(3);
  avec << a, a, a;

  Eigen::Matrix<double, Eigen::Dynamic, 1> bvec(3);
  bvec << 2.0, 2.0, 2.0;

  //Derivative types set through partials_return_type
  Eigen::Matrix<T_partials_return, Eigen::Dynamic, 1> dxvec(3);
  dxvec << 5.0, 5.0, 5.0;

  operands_and_partials<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>, 
                        Eigen::Matrix<double, Eigen::Dynamic, 1>> o(avec, bvec);

  if (!is_constant_struct<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>>::value)
    o.edge1_.partials_vec_[0] = dxvec;
  if (!is_constant_struct<Eigen::Matrix<double, Eigen::Dynamic, 1>>::value)
    o.edge2_.partials_vec_[0] = dxvec;

  o.build(-1.0);
}
```

Compilation fails with:
```
lib/eigen_3.3.3/Eigen/src/Core/functors/AssignmentFunctors.h:24:104: error: assigning to
      'double' from incompatible type 'const stan::math::var'
  ...void assignCoeff(DstScalar& a, const SrcScalar& b) const { a = b; }
```

andrjohns replied:
Looks like it's still edge2 causing the failure:
```
    o.edge2_.partials_vec_[0] = dxvec;
                                ^
1 error generated.
```

seantalts replied:
Ah, thanks for the improved test. It has been a while since I thought about mixed mode - I was forgetting that 1) we are only ever using a certain level of autodiff type at once (e.g. `fvar<var>` or `fvar<fvar<var>>` but never both) and that 2) a mixed mode o&p user would have vectors of vars to fake assign to the edges. I think having these new test cases in the PR would have helped a lot. Sorry this took me so long, but I think your solution is great! Couple of asks to get it in shape - can you standardize on one of either ViewElt / Op or T_part / T_op? I think I prefer the former (because the partials type might actually be a vector or matrix of T_part) but I think it's more important to just have them consistent (across all of the operands_and_partials.hpp files splayed out across the different directories). 

Second ask - can you add these tests and others to the PR? I think we either strongly like or require unit tests on new behavior like this, and as we've seen the tests are super helpful not just for validating but communication. 

I also would really like to see the original tests with their original values in the PR, if you don't mind. It should be easy enough to just include them as additional tests, I think?

I'm working on trying to refactor the 3 implementations of log_mix so that they are just 1 or 2 different ones... Seems like it should be possible but I don't want to ask you to do it without being sure.

seantalts replied:
Re: refactoring / generalizing - I think the best we can do is, for the row/column vector cases, additionally templating on `, int R, int C` and getting rid of the other version. Here is what that looks like: https://gist.github.com/seantalts/2185963296cde0e216250de3006525f3

andrjohns replied:
Not a problem at all! 
So the to-do list is:
* Replace ```T_op``` and ```T_part``` with ```0p``` and ```ViewElt```
* Add ```operands_and_partials``` tests for new functionality (```std::vector<std::vector<T>>```)
* Add tests to ```mix/mat/meta/operands_and_partials``` for mixed-mode with autodiff types (```fvar<var>``` and ```fvar<fvar<var>>```) and containers of doubles
* Re-add original ```log_mix``` tests that were changed
* Condense different implementations of ```log_mix```

Anything I've missed/misunderstood?

seantalts replied:
Sounds good. That should be `Op` and not `0p` but I think that's what you meant.

I think bullet 2 and 3 can be the same thing, pretty much. The new o&p use-cases you are fixing in this PR are all mixed-mode, and hopefully the old tests will continue to test the existing use-cases sufficiently. If you see any gaps there, even if you find yourself wondering "what would happen if..." feel free to add a test for that case! In this way, tests can also serve as documentation.

And the last one, I'm not sure there's a performant & readable way to condense the `std::vector` inner container specialization with the Eigen inner container ones. So don't worry about that for now, we can always come back to it.

Thank you very much for bearing with me!

andrjohns commented on stan/math/rev/mat/meta/operands_and_partials.hpp:
```diff
@@ -73,11 +74,11 @@ class ops_partials_edge<double, Eigen::Matrix<var, R, C> > {
 // (i.e. nested containers)
-template <int R, int C>
-class ops_partials_edge<double, std::vector<Eigen::Matrix<var, R, C> > > {
+template <typename T_part, int R, int C>
```
Yep the rev version did not need to change, I was supposed to clean that up before submitting! Thanks for catching.

andrjohns commented on test/unit/math/fwd/mat/fun/log_mix_test.cpp:
```diff
@@ -4,149 +4,165 @@
 using stan::math::fvar;
 using stan::math::log_mix;
-
-TEST(AgradFwdMatrixLogMix, vector_fd) {
```
I wrote the original tests, and they weren't very comprehensive (or efficient). I extended them in this pull to better test all possible instantiations with less copy-and-paste repetition.

seantalts replied:
This is good. I guess I'm just wondering how I should be verifying that the old behavior isn't changing with this new feature?

andrjohns replied:
By making the tests more comprehensive, the idea was that if the old behaviour changes then the tests would fail. Both versions tested the same things: whether the correct log-probability and partial derivatives were returned, the new version just tests it for more combinations of inputs. If the new feature caused the wrong values/derivatives to be returned for the other instantiations then the tests would fail.

seantalts commented:
What line is failing?
On Fri, Apr 6, 2018 at 22:44 Andrew Johnson <notifications@github.com>
wrote:

> *@andrjohns* commented on this pull request.
> ------------------------------
>
> In stan/math/prim/mat/meta/operands_and_partials.hpp
> <https://github.com/stan-dev/math/pull/751#discussion_r179906369>:
>
> > @@ -15,14 +15,14 @@ namespace internal {
>  /* This class will be used for both multivariate (nested container)
>     operands_and_partials edges as well as for the univariate case.
>   */
> -template <int R, int C>
> -class ops_partials_edge<double, Eigen::Matrix<double, R, C>> {
> +template <typename T_op, typename T_part, int R, int C>
>
> Progress is good! Unfortunately your changes break the usage of
> partials_return_type.
>
> If changing the test case posted above to:
>
> #include <stan/math/mix/mat.hpp>
> #include <stan/math/prim/scal/meta/is_constant_struct.hpp>
> #include <gtest/gtest.h>
>
> TEST(MetaTraits, partials_type) {
>   using stan::math::fvar;
>   using stan::math::var;
>   using stan::is_constant_struct;
>   using stan::math::operands_and_partials;
>
>   typedef typename stan::partials_return_type<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>,
>                                               Eigen::Matrix<double, Eigen::Dynamic, 1>>::type
>       T_partials_return;
>
>   fvar<var> a(2.0, 2.0);
>
>   Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1> avec(3);
>   avec << a, a, a;
>
>   Eigen::Matrix<double, Eigen::Dynamic, 1> bvec(3);
>   bvec << 2.0, 2.0, 2.0;
>
>   //Derivative types set through partials_return_type
>   Eigen::Matrix<T_partials_return, Eigen::Dynamic, 1> dxvec(3);
>   dxvec << 5.0, 5.0, 5.0;
>
>   operands_and_partials<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>,
>                         Eigen::Matrix<double, Eigen::Dynamic, 1>> o(avec, bvec);
>
>   if (!is_constant_struct<Eigen::Matrix<fvar<var>, Eigen::Dynamic, 1>>::value)
>     o.edge1_.partials_vec_[0] = dxvec;
>   if (!is_constant_struct<Eigen::Matrix<double, Eigen::Dynamic, 1>>::value)
>     o.edge2_.partials_vec_[0] = dxvec;
>
>   o.build(-1.0);
> }
>
> Compilation fails with:
>
> lib/eigen_3.3.3/Eigen/src/Core/functors/AssignmentFunctors.h:24:104: error: assigning to
>       'double' from incompatible type 'const stan::math::var'
>   ...void assignCoeff(DstScalar& a, const SrcScalar& b) const { a = b; }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/751#discussion_r179906369>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7CJzxAyN8fk3iX5dSFVNuYYzfaF0ks5tmCfxgaJpZM4R7nuw>
> .
>

andrjohns commented:
@seantalts That should about cover everything. Still working through approaches for condensing the std::vector and eigen container versions, but it's not really worth holding up this pull for

seantalts commented:
Hey, looks great! I think we still need some mix mode tests for the new operands_and_partials features you've enabled, and to condense just the two Eigen vector implementations (the std::vector one can stay separate). Thanks!

andrjohns commented:
Ah almost missed it, I've condensed the two eigen templates. For the ```operands_and_partials``` mixed mode tests I've added a matrix of doubles to all current instantiations and added an ```std::vector<std::vector<fvar<var>>>``` and matrix of doubles instantiation test.

Is there anything else you think I should add?

seantalts commented:
Sorry, lost track of this while I was traveling. Looks great, thank you so much! Merging.
